opam-ci-check/bin/main.ml

```diff
@@ -140,35 +194,63 @@ let pkg_term =
   let info = Arg.info [] ~doc:"Package name + version" in
   Arg.required (Arg.pos 0 (Arg.some Arg.string) None info)
 
-let changed_pkgs_term =
-  let info =
-    Arg.info
-      [ "c"; "changed-packages" ]
-      ~doc:"List of changed package name + version"
+let package_specs_term =
+  let package_spec_conv =
+    let parse_package_spec s =
+      try
+        let parts = String.split_on_char ':' s in
+        let name_version = List.hd parts in
+        let pkg = OpamPackage.of_string name_version in
+        let attrs =
+          match List.tl parts |> String.concat ":" with
+          | "" -> []
+          | attrs -> String.split_on_char ',' attrs
+        in
+        (* Allow only named attributes *)
+        List.iter (fun x -> assert (String.contains x '=')) attrs;
+        let src =
+          List.find_opt (fun x -> String.starts_with ~prefix:"src=" x) attrs
+          |> Option.map (fun x -> String.sub x 4 (String.length x - 4))
+        in
+        let newly_published =
+          List.find_opt (fun x -> String.starts_with ~prefix:"new=" x) attrs
+          |> Option.map (fun x -> String.sub x 4 (String.length x - 4) = "true")
+        in
+        Ok { pkg; src; newly_published }
+      with _ ->
+        Error
+          (`Msg
+            "Invalid PACKAGE_SPEC format. Expected \
+             <name.version>[:src=<path>][,new=<true|false>]")
+    in
+    let print_package_spec fmt spec =
+      let open Format in
+      fprintf fmt "%s" (OpamPackage.to_string spec.pkg);
+      (match spec.src with
+      | Some path -> fprintf fmt ":src=%s" path
+      | None -> ());
+      match spec.newly_published with
+      | Some true -> fprintf fmt ",new=true"
+      | Some false -> fprintf fmt ",new=false"
+      | None -> ()
+    in
+    Arg.conv (parse_package_spec, print_package_spec)
   in
-  Arg.value (Arg.opt (Arg.list Arg.string) [] info)
-
-let newly_published_pkgs_term =
   let info =
-    Arg.info [ "n"; "newly-published" ]
-      ~doc:"List of newly published package name + version"
-  in
-  Arg.value (Arg.opt (Arg.list Arg.string) [] info)
-
-let packages_term =
-  let create_term changed_pkgs newly_published_pkgs =
-    if changed_pkgs = [] && newly_published_pkgs = [] then
-      `Error
-        ( false,
-          "You must provide at least one changed or newly published package." )
-    else `Ok (changed_pkgs, newly_published_pkgs)
+    Arg.info []
+      ~doc:
+        "List of package specifications (format: \
+         <name.version>[:src=<path>][,new=<true|false>]). If [src] is not \
+         specified, the sources are downloaded from the source URL. If [new] \
+         is not specified, it is inferred from the opam repository."
```

Comment:
This seems like it will solve the problem we are currently facing in CLI flexibility, but I am also a bit nervous about introducing a new ad hoc specification format.

If we need to provide rich, structured data like this, an alternative might be to just use s-expressions (we already depend on Sexplib any how :smile:). E.g.,

```
opam-ci-check lint p.0.1.0 '(q.0.1.1 (src some/dir))' '(r.1.0 new)' '(s.1.1 (src a/dir) new)'
```

I don't think this is preferable, but the benefit IMO is that doesn't require inventing new format for compound data. But I am also don't want to waste time bikeshedding on a CLI input format that likely will not be used much by people (because of the upstreaming plans).

Perhaps this worth a short sync chat on the tradeoffs?